src/models.rs

```diff
@@ -62,6 +62,19 @@ pub(crate) trait StepCommon<'s> {
     /// Like [`Self::location()`], except with the step's `name`
     /// key as the final path component if present.
     fn location_with_name(&self) -> SymbolicLocation<'s>;
+
+    /// Returns the document which contains this step.
+    fn document(&self) -> &'s yamlpath::Document;
+}
+
+pub(crate) trait AsDocument<'a, 'doc> {
+    fn as_document(&'a self) -> &'doc yamlpath::Document;
+}
+
+impl<'a, 'doc, T: StepCommon<'doc>> AsDocument<'a, 'doc> for T {
+    fn as_document(&'a self) -> &'doc yamlpath::Document {
+        self.document()
+    }
 }
```

Comment:
This PR's diff is now a mess thanks to me renaming things, but this is the core of the change: we have an `AsDocument` trait that `build(...)` now takes.

(`AsDocument` has a slightly wonky lifetime set, to allow us to express both pre-existing lifetime patterns: things like `Step` where the document outlives its container, and things like `Workflow` where the document's lifetime is coextensive with its container.)